# Exploring Hacker News Posts

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

## Introduction

The first involves reading the data and removing the header

In [6]:
from csv import reader
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


## Removing Headers from a List of Lists

In [9]:
headers = hn[0] # extract the 1st row of data
hn = hn[1:]    # remove the 1st row from hn
print(headers)
print(hn[:5])  

['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']
[['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12'], ['10482257', 'Title II kills investment? Comcast and other ISPs are now spending more', 'http:

We can see above that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. Let's start by exploring the number of comments for each type of post.

## Extracting Ask HN and Show HN

First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [27]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts))
#print(ask_posts)
print(len(show_posts))
#print(show_posts)
print(len(other_posts))
#print(other_posts)
    

1744
1162
17193


## Calculating the Average Number of Comments in Ask HN and Show HN posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [28]:
# Calculate the average number of comments `Ask HN` posts receive.
total_ask_comments = 0

for comment in ask_posts:
    ask_comment = int(comment[4])
    total_ask_comments += ask_comment

avg_ask_comments = total_ask_comments/ len(ask_posts)
print(avg_ask_comments)   

14.038417431192661


In [30]:
total_show_comments = 0

for comment in show_posts:
    show_comment = int(comment[4])
    total_show_comments += show_comment

avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

## Finding the Amount of Ask Posts and Comments by Hour created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [47]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list = []
for post in ask_posts:
    result_list.append([post[6], int(post[4])])


counts_by_hour = {}
comments_by_hour = {}
date_format = '%m/%d/%Y %H:%M'

for row in result_list:
    date = row[0]
    comment = row[1]
    time_hour = dt.datetime.strptime(date, date_format).strftime("%H")
    if time_hour in counts_by_hour:
        counts_by_hour[time_hour] += 1
        comments_by_hour[time_hour] += comment
    else:
        counts_by_hour[time_hour] = 1
        comments_by_hour[time_hour] = comment
    
for key, value in comments_by_hour.items():
    print(key, ":", value)
print("-----------------------------------------------")
for key, value in counts_by_hour.items():
    print(key, ":", value)


07 : 267
23 : 543
11 : 641
05 : 464
12 : 687
04 : 337
20 : 1722
09 : 251
22 : 479
17 : 1146
18 : 1439
19 : 1188
16 : 1814
01 : 683
00 : 447
15 : 4477
13 : 1253
03 : 421
08 : 492
10 : 793
06 : 397
02 : 1381
21 : 1745
14 : 1416
-----------------------------------------------
07 : 34
23 : 68
11 : 58
05 : 46
12 : 73
04 : 47
20 : 80
09 : 45
22 : 71
17 : 100
18 : 109
19 : 110
16 : 108
01 : 60
00 : 55
15 : 116
13 : 85
03 : 54
08 : 48
10 : 59
06 : 44
02 : 58
21 : 109
14 : 107


## Calculating the average number of comments for Ask HN Posts by hour

In [52]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour] ])

avg_by_hour

[['07', 7.852941176470588],
 ['23', 7.985294117647059],
 ['11', 11.051724137931034],
 ['05', 10.08695652173913],
 ['12', 9.41095890410959],
 ['04', 7.170212765957447],
 ['20', 21.525],
 ['09', 5.5777777777777775],
 ['22', 6.746478873239437],
 ['17', 11.46],
 ['18', 13.20183486238532],
 ['19', 10.8],
 ['16', 16.796296296296298],
 ['01', 11.383333333333333],
 ['00', 8.127272727272727],
 ['15', 38.5948275862069],
 ['13', 14.741176470588234],
 ['03', 7.796296296296297],
 ['08', 10.25],
 ['10', 13.440677966101696],
 ['06', 9.022727272727273],
 ['02', 23.810344827586206],
 ['21', 16.009174311926607],
 ['14', 13.233644859813085]]

Although we now have the results we need, this format makes it hard to identify the hours with the highest values. So we are going to finish our analysis by sorting the above created list of list and printing the five highest values in a format that is easier to read

## Sorting and Printing Values from avg_by_hour list of list

In [56]:
swap_avg_by_hour = []
for each_row in avg_by_hour:
    swap_avg_by_hour.append([each_row[1], each_row[0]])
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Post Comments")
print("----------------------------------------------")
for avg,hr in sorted_swap[:5]:
    print('{}: {:.2f} average comments per post'.format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"),avg))
    

Top 5 Hours for Ask Post Comments
----------------------------------------------
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.